storage/local/storage.go

```diff
@@ -616,6 +609,54 @@ func (s *memorySeriesStorage) Append(sample *model.Sample) {
 	s.incNumChunksToPersist(completedChunksCount)
 }
 
+// Throttled implements Storage.
+func (s *memorySeriesStorage) Throttled() bool {
+	if s.getNumChunksToPersist() > s.maxChunksToPersist ||
+		float64(atomic.LoadInt64(&numMemChunks)) > float64(s.maxMemoryChunks)*toleranceFactorMemChunks {
+		s.throttled <- struct{}{}
+		return true
+	}
+	return false
+}
+
+// logThrottling handles logging of throttled events and has to be started as a
+// goroutine. It stops once s.loopStopping is closed.
+//
+// Logging strategy: Whenever Throttle() is called and returns true, an signal
+// is sent to s.throttled. If that happens for the first time, an Error is
+// logged that the storage is now throttled. As long as signals continues to be
+// sent via s.throttled at least once per minute, nothing else is logged. Once
+// no signal has arrived for a minute, an Info is logged that the storage is not
+// throttled anymore. This resets things to the initial state, i.e. once a
+// signal arrives again, the Error will be logged again.
+func (s *memorySeriesStorage) logThrottling() {
+	timer := time.NewTimer(time.Minute)
+	timer.Stop()
+
+	for {
+		select {
+		case <-s.throttled:
+			if !timer.Reset(time.Minute) {
+				log.
+					With("chunksToPersist", s.getNumChunksToPersist()).
+					With("maxChunksToPersist", s.maxChunksToPersist).
+					With("memoryChunks", atomic.LoadInt64(&numMemChunks)).
+					With("maxToleratedMemChunks", int(float64(s.maxMemoryChunks)*toleranceFactorMemChunks)).
+					Error("Storage is throttled. Scrapes and rule evaluations will be skipped.")
```

Comment:
I like the idea with `select` to skip if the channel has already been sent to. (But note that we wouldn't try to send on every call of Throttle, but only _if_ we are actually throttled, in which case we would not do a lot of harm by waiting as the caller will not start a scrape anyway.)

I thought long about not having the "stopping" log (following exactly your line of argument) and came to the conclusion it would be more confusing to the users to leave it out. Note that the message is _not_ repeated every minute. I invested some brain efforts to make it like the message for "rushed mode".

Note that the message will only be flapping if Throttle() is called less than once per minute. Which would hardly cause an overload of the storage in the first place. (Even if we are on a weak machine that can only ingest 10k samples/s, we needed scrapes that contain 600,000 metrics each to trigger that threshold with one call per minute.) Yes, in that pathological situation, you would get a flapping message, but you would probably have other problems, too.